lib/Sema/LookupVisibleDecls.cpp

```diff
@@ -1171,167 +1171,131 @@ static void lookupVisibleMemberDecls(
-        namelookup::FindLocalVal(SM, Loc, Consumer).checkValueDecl(
-          const_cast<ParamDecl *>(P), DeclVisibilityKind::FunctionParameter);
-      }
+namespace {
+class ASTScopeVisibleDeclConsumer
+    : public swift::namelookup::AbstractASTScopeDeclConsumer {
+  SourceLoc LookupLoc;
+  VisibleDeclConsumer &BaseConsumer;
 
-      namelookup::FindLocalVal(SM, Loc, Consumer).checkParameterList(
-        AFD->getParameters());
+  const DeclContext *const InnermostTypeDC;
+  DeclContext *SelfDC = nullptr;
 
-      GenericParams = AFD->getGenericParams();
+public:
+  ASTScopeVisibleDeclConsumer(SourceLoc Loc, const DeclContext *DC,
+                              VisibleDeclConsumer &BaseConsumer)
+      : LookupLoc(Loc), BaseConsumer(BaseConsumer),
+        InnermostTypeDC(DC->getInnermostTypeContext()) {}
 
-      if (AFD->getDeclContext()->isTypeContext()) {
-        ExtendedType = AFD->getDeclContext()->getSelfTypeInContext();
-        DC = DC->getParentForLookup();
+private:
+  void foundDecl(ValueDecl *VD) {
+    auto Kind = DeclVisibilityKind::LocalDecl;
+    if (isa<ParamDecl>(VD))
+      Kind = DeclVisibilityKind::FunctionParameter;
+    if (auto *GP = dyn_cast<GenericTypeParamDecl>(VD)) {
+      Kind = DeclVisibilityKind::GenericParameter;
+      // Generic param for 'some' parameter type is not "visible".
+      if (GP->isOpaqueType())
+        return;
+    }
+    BaseConsumer.foundDecl(VD, Kind);
+  }
 
-        if (auto *FD = dyn_cast<FuncDecl>(AFD))
-          if (FD->isStatic())
-            LS = LS.withOnMetatype();
-      }
-    } else if (auto CE = dyn_cast<ClosureExpr>(DC)) {
-      if (Loc.isValid()) {
-        namelookup::FindLocalVal(SM, Loc, Consumer).visit(CE->getBody());
-        if (auto P = CE->getParameters()) {
-          namelookup::FindLocalVal(SM, Loc, Consumer).checkParameterList(P);
+  bool consume(ArrayRef<ValueDecl *> values,
+               NullablePtr<DeclContext> baseDC) override {
+    for (auto *VD : values) {
+      if (auto *var = dyn_cast<VarDecl>(VD)) {
+        // If we have the 'self' parameter, make a note of the DeclContext
+        // where it exists.
+        if (var->isSelfParameter())
+          SelfDC = var->getDeclContext();
+
+        if (var->getPropertyWrapperBackingProperty()) {
+          // FIXME: This is currently required to set the interface type of the
+          // auxiliary variables (unless 'var' is a closure param).
+          (void)var->getPropertyWrapperBackingPropertyType();
         }
+        var->visitAuxiliaryDecls(
+            [&](VarDecl *auxVar) { foundDecl(auxVar); });
       }
-    } else if (auto ED = dyn_cast<ExtensionDecl>(DC)) {
-      ExtendedType = ED->getSelfTypeInContext();
-    } else if (auto ND = dyn_cast<NominalTypeDecl>(DC)) {
-      ExtendedType = ND->getSelfTypeInContext();
+      VD->visitAuxiliaryDecls([&](Decl *auxDecl) {
+        if (auto *auxValue = dyn_cast<ValueDecl>(auxDecl))
+          foundDecl(auxValue);
+      });
```

Comment:
Seems like ASTScopeDeclConsumerForLocalLookup currently just doesn't handle peer macros 🤔

Reply:
The goal is to eventually rewrite ASTScope to look at swift-syntax only, so semantic members should be introduced by clients (but we can factor out a common client for everyone to use that also visits semantic local members).

Reply:
Okay,

Previous `FindLocalVal` implementation handling aux decls was recursive.
```
  std::function<void(Decl *)> visitDecl;
  visitDecl = [&](Decl *D) {
    if (auto *VD = dyn_cast<ValueDecl>(D))
      checkValueDecl(VD, DeclVisibilityKind::LocalVariable);
    D->visitAuxiliaryDecls(visitDecl);
  };
  ...
  visitDecl(D);
```
This was to handle multi level macro expansion (e.g. `peer` macro expand to a decl with another `peer` macro) I _think_ we need similar treatment.

Reply:
After speaking offline, updated to match ASTScopeDeclConsumerForLocalLookup, since peer decls for local decls should not be included in the lookup result.